### PR TITLE
Access data flow

### PR DESCRIPTION
**Pull request by injms**
### Context
Registers has changed the way data is accessed, removing the need for the API key. This new data flow should mirror Conor's design and Mark's content, as found on the [ticket](https://trello.com/c/UtIsuXtd/2739-implement-new-access-the-data-flow)

### Changes proposed in this pull request
 - On a register's page, the 'Access the data' no longer an accordion. It now links to the access the data flow.
 - Access the data is now a 3 step process: 
   1. API or download
   2. Help us improve
   3. API details or Download page.
 - Once a user has gone through the access the data process once, they won't see step 2 for two weeks.
 - The 'Help us improve' form is optional, and can be skipped by clicking the 'Skip this step' or 'Prefer not to say' links
 - A user can then choose to sign up for updates on the signup page; successful submission leads to a thank you page
 - The email address in the 'signup for updates' form gets added to a list in MailChimp. 
 - Both of the forms have event tracking on them
   - The 'access the data' form fires events into GA rather than into a backend. The government departments input can only select from a predefined list.
   - The 'signup for updates' form only fires an event on submission; we only look at the domain of the email address. 
 - This uses the new GOV.UK design system, which requires NPM to pull it in. (This meant changes to the Travis build, CloudFoundry ignore, gitignore, Dockerfile,assets.rb; and inclusion of package.json and package-lock.json).

### Guidance to review
 - Cliff has reviewed the Analytics events.

### CODE REVIEW

**Comment by MatMoore** [comment]
Can you squash together the WIP commits into a smaller number of changes? Love the ascii art though 😄

File: app/views/download/choose_access.html.haml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+          %li use our quick start guide to get set up
+          %li get notifications about updates to our service, like downtime and breaking changes
+        = link_to 'Use the API', @next_step_api, { class: 'js--ga-watch-click', id: "label-API" }
+
+      .highlight-box
+        %h3.heading-medium Download
+        %p
+          You can download the data in each register.
+        %ul.list.list-bullet
+          %li get the data as JSON, CSV, and ODS
```
Comment (gidsg) [inline app/views/download/choose_access.html.haml]: ODS is not supported yet, we should remove this reference.

File: app/views/download/help_improve.html.haml [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      }
+
+      var skipStepLinks = document.querySelectorAll('.js--ga-watch-click');
+      for (var j = 0; j < skipStepLinks.length; j++) {
+        skipStepLinks[j]
+          .addEventListener('click', function(event) {
+            ga('send', {
+              hitType: 'event',
+              eventCategory: 'Form',
+              eventAction: 'Click - I\'d prefer not to say',
```
Comment (gidsg) [inline app/views/download/help_improve.html.haml]: can we use double quotes rather than escaping?

File: app/views/registers/show.html.haml [hunk 3]
```diff
@@ -119,14 +119,11 @@
...
-                %h2.subsection__title Access the data
-              .subsection__content.js-subsection-content#access_section
-                %ul.list
-                  %li= link_to 'Create API key', api_users_path(register: @register.slug)
-                  %li Always access the most recent version of all registers through our API.
-                %p You can #{link_to 'download the latest data', register_download_index_path(@register.slug)} in this register as a CSV or JSON file.
+            .subsection--not-accordionised
+              .subsection__header--not-accordionised
+                %h2.subsection__title--not-accordionised 
+                  -# = link_to_choose_access
```
Comment (gidsg) [inline app/views/registers/show.html.haml]: I think this commented out line can be deleted?

File: config/routes.rb [hunk 4]
```diff
@@ -10,6 +10,14 @@
   get 'privacy-notice', to: 'pages#privacy_notice', as: 'privacy_notice'
   get 'cookies', to: 'pages#cookies', as: 'cookies'
 
+  resources :sign_up, only: %i[create index], path: 'sign-up-for-updates' do
+    # collection do
```
Comment (gidsg) [inline config/routes.rb]: we should delete this as it is commented out

File: package.json [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  "description": "The frontend for GOV.UK Registers.",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/openregister/registers-frontend.git"
+  },
+  "author": "https://github.com/orgs/alphagov/teams/gov-uk-registers",
+  "license": "UNLICENSED",
```
Comment (gidsg) [inline package.json]: the license of the repo is MIT.

File: app/assets/javascripts/application.js [hunk 6]
```diff
@@ -23,6 +23,9 @@
 //= require jquery-ui/widget
 //= require jquery-ui/sortable
 
+// GOV.UK Design System
+// = require all.js
```
[the remark this answers is not in the document]
Comment (injms) [reply app/assets/javascripts/application.js]: This isn't commented out, despite looking like it is. I've removed the space between `//` and `=` to make it clearer that this is the same as the previous `//= require` in the file.

File: .gitignore [hunk 7]
```diff
@@ -16,9 +16,12 @@
...
-# Ignore jet brains tools config
+# Ignore jet brains tools config.
 .idea
 *.iml
 
-# ignore outputs from travis
-/vendor
\ No newline at end of file
+# Ignore outputs from travis.
+/vendor
```
[the remark this answers is not in the document]
Comment (injms) [reply .gitignore]: Cheers - now removed from `.gitignore`.

File: README.md [hunk 8]
```diff
@@ -9,9 +9,16 @@ Frontend for [GOV.UK Registers](https://www.registers.service.gov.uk/)
 1. Install Ruby 2.5.0
 2. Install Rails 5.1.4
 3. Install PostgreSQL 9.6+
+4. Install Node 10+
```
[the remark this answers is not in the document]
Comment (injms) [reply README.md]: `.nvmrc` added with version 10.9.0

File: app/controllers/download_controller.rb [hunk 9]
```diff
@@ -3,16 +3,26 @@
 class DownloadController < ApplicationController
   include FormHelpers
   before_action :set_register
+  before_action :set_government_organisations, only: :new
   helper_method :government_orgs_local_authorities
 
   def index
-    @download = DownloadUser.new
+    @number_of_steps = cookies[:seen_help_us_improve_questions] ? 2 : 3
+    @custom_dimension = @register.slug.tr('-', ' ') + ' - download'
```
[the remark this answers is not in the document]
Comment (injms) [reply app/controllers/download_controller.rb]: Fixed on all occurrences in this file. I'll keep an eye out to see where else I can fix this too.

File: app/controllers/download_controller.rb [hunk 10]
```diff
@@ -38,6 +48,49 @@ def create
...
   def success; end
 
+  def choose_access
+    if cookies[:seen_help_us_improve_questions]
+      @number_of_steps = 2
+      @next_step_api = register_get_api_path(@register.slug)
+      @next_step_download = register_path(@register.slug) + '/download'
+    else
+      @number_of_steps = 3
+      @next_step_api = register_help_improve_api_path(@register.slug);
```
[the remark this answers is not in the document]
Comment (injms) [reply app/controllers/download_controller.rb]: I'll need to ask you about this to get my head around when we do and don't need to provide arguments in rails. A lot of magic going on here.

File: app/controllers/download_controller.rb [hunk 11]
```diff
@@ -38,6 +48,49 @@ def create
...
+      @next_step_download = register_help_improve_download_path(@register.slug)
+    end
+  end
+
+  def help_improve
+    @government_organisations = Register.find_by(slug: 'government-organisation')
+                                        .records
+                                        .where(entry_type: 'user')
+                                        .current
+    if request.fullpath.match?(/api$/)
```
[the remark this answers is not in the document]
Comment (injms) [reply app/controllers/download_controller.rb]: The two routes are to make it easier for Analytics. As discussed, I've tagged this with a TODO so we can come back to this and do it more elegantly.

**Comment by gidsg** [comment]
would be good to add a feature spec for access the data, as we are deleting the previous one.

File: app/models/help_us_improve_user.rb [hunk 12]
```diff
@@ -0,0 +1,5 @@
+class HelpUsImproveUser
```
Comment (MatMoore) [inline app/models/help_us_improve_user.rb]: Is this model actually used? It seems like we don't process this input on the backend, we just track this in GA... is that right?

File: app/controllers/download_controller.rb [hunk 13]
```diff
@@ -3,41 +3,73 @@
...
+  before_action :set_government_organisations, only: :new
   helper_method :government_orgs_local_authorities
 
   def index
-    @download = DownloadUser.new
+    @number_of_steps = cookies[:seen_help_us_improve_questions] ? 2 : 3
+    @custom_dimension = "#{@register.name} - download"
   end
 
   def create
```
[the remark this answers is not in the document]
Comment (injms) [reply app/controllers/download_controller.rb]: The `@custom_dimension` variable is for tagging any page views or events that get sent to GA - the wording was requested by Cliff.

File: app/controllers/download_controller.rb [hunk 14]
```diff
@@ -54,14 +86,5 @@ def set_register
...
-    params.require(:download_user).permit(
-      :email_gov,
-      :email_non_gov,
-      :department,
-      :non_gov_use_category,
-      :is_government,
-      :register
-    )
-  end
+  def download_user_params; end
```
[the remark this answers is not in the document]
Comment (injms) [reply app/controllers/download_controller.rb]: Removed and things still work - which is always a bonus.

File: config/routes.rb [hunk 15]
```diff
@@ -11,15 +11,29 @@
...
-    get '/download/success', to: 'download#success', as: 'download_success'
+
     get '/download-json', to: 'download#download_json', as: 'download_json'
     get '/download-csv', to: 'download#download_csv', as: 'download_csv'
+
+    get '/choose-how-to-access', to: 'download#choose_access', as: 'choose_access'
+    get '/help-us-improve-the-api', to: 'download#help_improve', as: 'help_improve_api'
+    get '/help-us-improve', to: 'download#help_improve', as: 'help_improve_download'
+    get '/use-the-api', to: 'download#get_api', as: 'get_api'
+    post '/use-the-api', to: 'download#post_api', as: 'post_api'
```
[the remark this answers is not in the document]
Comment (injms) [reply config/routes.rb]: This is used on the form on the 'Help us improve' page - it distinguish between users who have submitted the form as a `post` and users who have decided to skip the form as a `get`.

File: app/models/help_us_improve_user.rb (hunk 12 above)
Comment (injms) [reply app/models/help_us_improve_user.rb]: That's correct - I've now removed the unused model.

File: app/controllers/download_controller.rb [hunk 16]
```diff
@@ -3,41 +3,73 @@
...
 
   def index
-    @download = DownloadUser.new
+    @number_of_steps = cookies[:seen_help_us_improve_questions] ? 2 : 3
+    @custom_dimension = "#{@register.name} - download"
   end
 
   def create
-    @download = DownloadUser.new(download_user_params.merge!(register: params[:register_id]))
+    @number_of_steps = cookies[:seen_help_us_improve_questions] ? 2 : 3
```
[the remark this answers is not in the document]
Comment (injms) [reply app/controllers/download_controller.rb]: Good point.

File: README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (gidsg) [reply README.md]: Can we use the same node version used by design system?  Which is 8 LTS https://github.com/alphagov/govuk-design-system/blob/master/.nvmrc

**Pull request merged**